<?xml version="1.0" ?>

<launch>
	<!-- If mode = sim: Simulated AUV with MBES and control in loaded bathymetric map 
	If mode = gt: reproduce Hugin survey and construct map of real bathymetry --> 
    <arg name="mode"                            default="sim"/>
    <arg name="namespace"                       default="lolo"/>
    <arg name="debug"                           default="0"/>
    <arg name="dataset"                         default="lost_targets"/>
    <arg name="app"                             default="ipp"/>

    <!-- MBES sensor -->
	<arg name="n_beams_mbes"                    default="100"/>	
	<arg name="mbes_open_angle"                 value="2.250" />
  
    <!-- Coordinates of odom frame wrt the map frame. By default, the AUV start the mission here -->
    <arg name="x"                               default="0"/>
    <arg name="y"                               default="0"/>
    <arg name="z"                               default="0.0"/>
    <arg name="roll"                            default="0.0"/>
    <arg name="pitch"                           default="0.0"/>
    <arg name="yaw"                             default="0.0"/>

    <arg name="map_frame"                       default="map"/>
    <arg name="odom_frame"                      default="$(arg namespace)/odom"/>
    <arg name="utm_frame"                      default="utm"/>
    <arg name="odometry_topic"                  default="/lolo/ipp/odom"/>
    <arg name="mbes_pings_topic"                default="/lolo/sensors/mbes/pc2_bathymetry"/>

    <!-- RBPF params  -->
    <arg name="particle_count"                  default="1"/>
    <arg name="num_particle_handlers"           default="1"/> 
    <arg name="survey_finished_top"             default="/$(arg mode)/survey_finished"/>

    <!-- API for SVGP particles -->
    <arg name="gp_ip_topic"                     default="/$(arg namespace)/gp/inducing_points"/>
    <arg name="gp_mb_server"                    default="/$(arg namespace)/gp/minibatch_server"/>
    <arg name="manipulate_gp_server"            default="/particle_0/$(arg namespace)/gp/manipulate_server"/>
    <arg name="gp_resample_top"                 default="/$(arg namespace)/gp/resample_top"/>
    <arg name="results_path"                    default="/home/sam/ipp_ws/src/UWExploration/planning/ipp/data/"/>   
    <arg name="num_particles_per_handler"       default="1"/>      
    <arg name="svgp_minibatch_size"   default="1000" />

    <!-- Planner parameters -->
    <!-- Planner parameters -->
    <arg name="planner_type"                    default="opt"/>
    <arg name="turning_radius"                  default="20"/>
    <arg name="corner_topic"                    default="/$(arg namespace)/corners"/>
    <arg name="path_topic"                      default="/$(arg namespace)/waypoints"/>
    <arg name="planner_req_topic"               default="/$(arg namespace)/planner_req"/>
    <arg name="swath_width"                     default="18"/>
    <arg name="swath_overlap"                   default="0.2"/>
    <arg name="training_rate"                   default="10"/>
    <arg name="wp_resolution"                   default="20"/>
    <arg name="path_nbr_samples"                default="7" />
    <arg name="voxel_size"                      default="3" />
    <arg name="wp_sample_interval"              default="1" />
    <arg name="beta"                            default="9" />

    <!-- Planner constraints -->
    <arg name="bound_left"                      default="592"/>
    <arg name="bound_right"                     default="821"/>
    <arg name="bound_up"                        default="-179"/>
    <arg name="bound_down"                      default="-457"/>
    <arg name="horizon_distance"                default="100"/>
    <arg name="border_margin"                   default="25"/>
    <arg name="max_mission_time"                default="3600"/>
    <arg name="vehicle_velocity"                default="0.8"/>

    <!-- If plotting decision support material should be enabled-->
    <arg name="visualize"                       default="1"/>

    <!-- Run IPP node -->
    <node type="planner_node.py" pkg="ipp" name="AUV_path_planning_node" output="screen"> 
        <param name="gp_ip_topic"               value="/$(arg namespace)/gp/inducing_points"/>
        <param name="inducing_points_top"       value="$(arg gp_ip_topic)"/>   
        <param name="gp_mb_server"              value="/$(arg namespace)/gp/minibatch_server"/>
        <param name="manipulate_gp_server"      value="/particle_0/$(arg namespace)/gp/manipulate_server"/>
        <param name="gp_resample_top"           value="/$(arg namespace)/gp/resample_top"/>
        <param name="gp_resampling_top"         value="/$(arg namespace)/gp/resample_top"/>
        <param name="results_path"              value="/home/alex/catkin_ws/src/UWExploration/planning/ipp/data/"/>
        <param name="sample_gp_server"          value="/$(arg namespace)/gp/sample_server"/>
        <param name="minibatch_gp_server"       value="$(arg gp_mb_server)"/>     
        <param name="svgp_num_ind_points"       value="400"/>     
        <param name="svgp_minibatch_size"       value="1000"/>     
        <param name="svgp_learning_rate"        value="1e-1"/>     
        <param name="svgp_rtol"                 value="1e-3"/>  
        <param name="svgp_verbose"              value="False"/>    
        <param name="svgp_auto_stop"            value="False"/>     
        <param name="num_particles_per_handler" value="$(arg num_particles_per_handler)"/>     
        <param name="results_path"              value="$(arg results_path)"/> 
        <param name="particle_enable_lc"        value="/$(arg namespace)/enable_lc"/> 
        <param name="planner_type"              value="$(arg planner_type)" />
        <param name="turning_radius"            value="$(arg turning_radius)" />
        <param name="corner_topic"              value="$(arg corner_topic)" />
        <param name="path_topic"                value="$(arg path_topic)" />
        <param name="planner_req_topic"         value="$(arg planner_req_topic)" />
        <param name="odom_topic"                value="$(arg odometry_topic)" />
        <param name="beta"                      value="$(arg beta)" />
        <param name="swath_width"               value="$(arg swath_width)" />
        <param name="swath_overlap"             value="$(arg swath_overlap)" />
        <param name="x"                         value="$(arg x)"/>
        <param name="y"                         value="$(arg y)"/>
        <param name="z"                         value="$(arg z)"/>
        <param name="roll"                      value="$(arg roll)"/>
        <param name="pitch"                     value="$(arg pitch)"/>
        <param name="yaw"                       value="$(arg yaw)"/>
        <param name="bound_left"                value="$(arg bound_left)" />
        <param name="bound_right"               value="$(arg bound_right)" />
        <param name="bound_up"                  value="$(arg bound_up)" />
        <param name="bound_down"                value="$(arg bound_down)" />
        <param name="border_margin"             value="$(arg border_margin)" />
        <param name="horizon_distance"          value="$(arg horizon_distance)" />
        <param name="wp_sample_interval"        value="$(arg wp_sample_interval)" />
        <param name="wp_resolution"             value="$(arg wp_resolution)" />
        <param name="voxel_size"                value="$(arg voxel_size)" />
        <param name="path_nbr_samples"          value="$(arg path_nbr_samples)" />
        <param name="max_mission_time"          value="$(arg max_mission_time)"/>
        <param name="vehicle_velocity"          value="$(arg vehicle_velocity)"/>
        <param name="train_rate"    type="int"  value="$(arg training_rate)" />
        <param name="svgp_n_window" type="int"  value="400"/>     
        <param name="n_beams_mbes"  type="int"  value="$(arg n_beams_mbes)" />    
    </node>

    <!-- <node type="particles_launcher.py" pkg="rbpf_slam" name="particle_launcher" output="screen">
        <param name="namespace" value="$(arg namespace)"/>  
        <param name="n_beams_mbes" value="$(arg n_beams_mbes)"/>  
        <param name="results_path" value="$(arg results_path)" />
        <param name="svgp_minibatch_size" value="$(arg svgp_minibatch_size)"/>     
        <param name="num_particle_handlers" value="$(arg num_particle_handlers)"/>     
        <param name="num_particles_per_handler" value="$(eval arg('particle_count') / arg('num_particle_handlers'))"/>     
        <param name="particle_launch_file" value="$(find rbpf_slam)/launch/particle.launch"/>     
    </node> -->

    <node type="ins_2_dr.py" pkg="lolo_ipp_interface" name="ins_2_dr" output="screen">
        <param name="map_frame"                 value="$(arg map_frame)" />
        <param name="odom_frame"                 value="$(arg odom_frame)" />
        <param name="utm_frame"                 value="$(arg utm_frame)" />
    </node>
    
    <!-- Run uncertainty management node -->
    <node type="auv_ui_online.py" pkg="uncert_management" name="auv_ui_online" output="screen">
        <param name="map_frame"                 value="$(arg map_frame)" />
        <param name="odom_frame"                value="$(arg odom_frame)" />
        <param name="base_link"                 value="$(arg namespace)/base_link" />
        <param name="mbes_link"                 value="$(arg namespace)/mbes_link" />
        <param name="odometry_topic"            value="/lolo/core/odom" />
        <param name="mbes_pings_topic"          value="$(arg mbes_pings_topic)" />
        <param name="background_img_path"       value="$(find uw_tests)/datasets/$(arg dataset)/default_real_mean_depth.png" />       
        <param name="dataset"                   value="$(arg dataset)" />       
        <param name="visualization_period"      value="0.5" />       
        <param name="survey_finished_top"       value="$(arg survey_finished_top)" />
        <param name="minibatch_gp_server"       value="$(arg gp_mb_server)"/>    
        <param name="inducing_points_top"       value="$(arg gp_ip_topic)"/>     
        <param name="corners_topic"                value="/$(arg namespace)/corners" />
        <param name="gp_resampling_top"         value="$(arg gp_resample_top)" />
        <param name="manipulate_gp_server"      value="$(arg manipulate_gp_server)" />
        <param name="synch_topic"               value="/$(arg app)/synch" />
    </node>


    <node type="w2w_client.py" pkg="lolo_ipp_interface" name="w2w_client" output="screen">
        <param name="path_planner_as"  value="/$(arg namespace)/ctrl/goto_waypoint" />
        <param name="path_topic"  value="$(arg path_topic)" />
        <param name="map_frame"  value="map" />
        <param name="wp_topic"  value="/navigation/lc_wp" />
        <!-- <param name="relocalize_topic"  value="/$(arg namespace)/pause_planner"/> -->
        <param name="synch_topic" value="/$(arg app)/synch" />
        <param name="planner_req_topic" value="/$(arg namespace)/planner_req" />
    </node>

    <!-- Run simulation environment -->
    <!-- <include file="$(find auv_model)/launch/auv_environment.launch" >
        <arg name="mode"                        default="sim"/>
        <arg name="namespace"                   default="hugin_0"/>
        <arg name="dataset"                     default="lost_targets"/>
        <arg name="x"                           default="0"/>
        <arg name="y"                           default="0"/>
    </include> -->

    <!-- datafile-->
    <include file="$(find auv_model)/launch/auv_env_aux.launch" >
        <arg name="dataset"                     default="asko"/>
        <arg name="namespace"                     default="$(arg namespace)"/>
    </include>

    <!-- If we are visualizing, plot decision material -->
    <!-- <group if="$(eval arg('visualize') == 1)">
        <node type="VisualizationClass.py" pkg="ipp" name="Visualization_node" output="screen">
            <param name="bound_left"                value="$(arg bound_left)" />
            <param name="bound_right"               value="$(arg bound_right)" />
            <param name="bound_up"                  value="$(arg bound_up)" />
            <param name="bound_down"                value="$(arg bound_down)" />    
            <param name="beta"                      value="$(arg beta)" />
        </node>
    </group> -->

</launch>

